source/slang/slang-ir-hlsl-legalize.cpp

```diff
@@ -0,0 +1,93 @@
+{
+    for(auto child : inst->getChildren())
+    {
+        switch(child->getOp())
+        {
+        case kIROp_Block:
+        {
+            searchChildrenForForceVarIntoStructTemporarily(module, child);
+            break;
+        }
+        case kIROp_Call:
+        {
+            auto call = as<IRCall>(child);
+            for(UInt i = 0; i < call->getArgCount(); i++)
+            {
+                auto arg = call->getArg(i);
+                if(arg->getOp() != kIROp_ForceVarIntoStructTemporarily)
+                    continue;
+                auto forceStructArg = arg->getOperand(0);
+                auto forceStructBaseType = as<IRType>(forceStructArg->getDataType()->getOperand(0));
+                if(forceStructBaseType->getOp() == kIROp_StructType)
+                {
+                    arg->replaceUsesWith(arg->getOperand(0));
+                    arg->removeAndDeallocate();
+                    continue;
+                }
+
+                // Replace all non struct inputs with following logic
+                // {`Non-struct var` to `struct var` copy }
+                // {Use `struct var` made as parameter to kIROp_Call}
+                // if(parameterIsPointerLikeTypeWhichMayHaveSideEffectsOnParameter) {`Struct var` to `non-struct` var copy}
+                IRBuilder builder(call);
+
+                builder.setInsertBefore(call->getCallee());
+                auto structType = builder.createStructType();
+                StringBuilder structName;
+                builder.addNameHintDecoration(structType, UnownedStringSlice("ForceVarIntoStructTemporarily_t"));
+
+                auto elementBufferKey = builder.createStructKey();
+                builder.addNameHintDecoration(elementBufferKey, UnownedStringSlice("data"));
+                auto _dataField = builder.createStructField(structType, elementBufferKey, forceStructBaseType);
+
+                builder.setInsertBefore(call);
+                auto structVar = builder.emitVar(structType);
+                builder.addNameHintDecoration(structVar, UnownedStringSlice("forceVarIntoStructTemporarily"));
+                builder.emitStore(
+                    builder.emitFieldAddress(builder.getPtrType(_dataField->getFieldType()), structVar, _dataField->getKey()),
+                    builder.emitLoad(forceStructArg));
+
+                arg->replaceUsesWith(structVar);
+                arg->removeAndDeallocate();
+
+                auto argType = call->getCallee()->getDataType()->getOperand(i+1);
+                if (!isPtrLikeOrHandleType(argType))
+                    continue;
+
+                builder.setInsertAfter(call);
+                builder.emitStore(
+                    forceStructArg,
+                    builder.emitFieldAddress(builder.getPtrType(_dataField->getFieldType()), structVar, _dataField->getKey()));
+            }
+            break;
+        }
+        }
+    }
+}
+
+void legalizeNonStructParameterToStructAndBackHLSL(IRModule* module)
```

Comment:
I am not sure why the function name has "AndBackHLSL".
What does it mean by it?
I think it could be just "legalizeNonStructParameterToStructForHLSL()".

Reply:
I will remove `AndBack`, in hindsight it makes little sense

it was supposed to imply "we handle a copy after transformation if required", but this makes little sense.